---
title: "Statistical Analysis of Methylation Data"
subtitle: ""
author: "Prof. Dr. Jochen Kruppa-Scheetz"
date: "13. May 2024"
format:
  revealjs: 
    logo: logo_osna.png
    css: logo.css
    preview-links: auto
    footer: "[Openbook Bio Data Science | The definitive guide](https://jkruppa.github.io/)"
    theme: simple
    slide-number: c/t
    chalkboard: 
      theme: whiteboard
---

## Why methylation analysis?

- Methylation is a crucial epigenetic modification that **regulates** gene expression without altering the DNA sequence.

- Abnormal methylation patterns are associated with numerous **diseases**, including cancer, autoimmune diseases, and neurological disorders.

-  In **cancer**, certain genes become abnormally methylated, leading to oncogene activation or tumor suppressor gene silencing.

::: .aside
*More information: [DNA Methylation and Its Basic Function](https://www.nature.com/articles/npp2012112)*
:::

## Why methylation analysis?

- Methylation plays a key in **development** by controlling the expression of genes critical for cell differentiation and organ formation. 

-  Methylation patterns change with **age** and in response to environmental factors, such as diet, toxins, and stress.

-  As methylation can be influenced by genetic and environmental factors, its analysis is essential for **personalized medicine**. 

::: .aside
*More information: [DNA Methylation and Its Basic Function](https://www.nature.com/articles/npp2012112)*
:::

## Dutch famine of 1944–1945 

![](img/hongerwinter.jpg){fig-align="center" width=80%}

## Dutch famine of 1944–1945

- Food caloric intake for many Dutch people dropped to fewer than 1000 calories per day. 
- Provided strong evidence for the concept of "fetal programming". 
- Changes in methylation patterns have been implicated in subsequent health outcomes. 
- Findings suggest that severe malnutrition during critical periods of development can lead to epigenetic marks that persist throughout life.

# Methylation is epigenetic 

::: .aside
*"on top of" or "in addition to" genetic*
:::

## Landscape of methylation

![](img/DNAme_landscape.png){fig-align="center" width=100%}

## Landscape of methylation in development

![](img/DNA_methylation_reprogramming.png){fig-align="center" width=100%}

## How can I help? 

![](img/mr-dna.png){fig-align="center" width="50%"}

. . .

1) Differential analysis
2) Prediction model


## Differential analysis

$$
Y \sim X + C
$$

- $Y$ is our outcome 
- $X$ is our factor with two levels. Our two or more treatment groups
- $C$ is our confunder effect of gender or age

**Aim**: To find differences in $Y$ between two or more treatment groups

## Prediction model

$$
Y \sim X_1 + ... + X_p
$$

- $Y$ is our binary outcome ($0/1$) 
- $X_1, ..., X_p$ are the normalised or standardised features

**Aim**: To predict $Y$ by features 


## Data

### Microarray-based methods

- Illumina HumanMethylation450 (450k) 
- Illumina Infinium MethylationEPIC BeadChip targets over 850,000 CpG sites
- We have idat files  

### High-throughput sequencing

- Bisulfite sequencing (BS-Seq) data
- We have fastq or sam/bam files


## Illumina array 

- Methylated intensity (denoted by M) and unmethylated intensity (denoted by U)
- $\beta\mbox{-values} = \cfrac{M}{M+U}$
- $\mbox{M-values} = log2\left(\cfrac{M}{U}\right)$

. . .

**Caution** 

We will use for different aspects the same letters. This might be confusing...

## Any experience?

*Gesa M. Richter and Jochen Kruppa, et al.*

- [A combined epigenome- and transcriptome-wide association study of the oral masticatory mucosa assigns CYP1B1 a central role for epithelial health in smokers](https://link.springer.com/article/10.1186/s13148-019-0697-y)

- [Epigenetic adaptations of the masticatory mucosa to periodontal inflammation](https://link.springer.com/article/10.1186/s13148-021-01190-7)

## Further reading {.smaller}

- [DNA Methylation and Its Basic Function](https://www.nature.com/articles/npp2012112)
- [Chapter 10 DNA methylation analysis using bisulfite sequencing data](https://compgenomr.github.io/book/bsseq.html) of the great book [Computational Genomics with R](https://compgenomr.github.io/book/)
- [A cross-package Bioconductor workflow for analysing methylation array data](https://www.bioconductor.org/packages/release/workflows/vignettes/methylationArrayAnalysis/inst/doc/methylationArrayAnalysis.html)
- [DNA Methylation: Bisulfite Sequencing Workflow](https://github.com/NBISweden/workshop-epigenomics-RTDs/blob/master/docs/content/tutorials/methylationSeq/Seq_Tutorial.rst)
- [Analysis of 450k data using `{minfi}`](https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html) and [The minfi User’s Guide](https://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.html)
- [`{methylKit}`: User Guide v1.2.4](https://bioconductor.riken.jp/packages/3.5/bioc/vignettes/methylKit/inst/doc/methylKit.html)
- [`{methylGSA}`: Gene Set Analysis for DNA Methylation Datasets](https://bioconductor.org/packages/release/bioc/vignettes/methylGSA/inst/doc/methylGSA-vignette.html)
- [`{BioMethyl}`: an R package for biological interpretation of DNA methylation data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6761945/)
- [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/)


## References {.smaller}

- Kruppa, J. *et al.* (2021). Estimands in epigenome-wide association studies. Clinical epigenetics, 13(1), 98.
- Maksimovic, J. *et al.* (2016). A cross-package Bioconductor workflow for analysing methylation array data. F1000Research, 5.
- Krueger, F. *et al.* (2011). Bismark: a flexible aligner and methylation caller for Bisulfite-Seq applications. bioinformatics, 27(11), 1571-1572.
- Aryee, M. J. *et al.* (2014). Minfi: a flexible and comprehensive Bioconductor package for the analysis of Infinium DNA methylation microarrays. Bioinformatics, 30(10), 1363-1369.
- Maksimovic, J. *et al.* (2021). Gene set enrichment analysis for genome-wide DNA methylation data. Genome Biology, 22(1), 173.
- Wang, Y. *et al.* (2019). BioMethyl: an R package for biological interpretation of DNA methylation data. Bioinformatics, 35(19), 3635-3641.

## Sources {.smaller}







